rootmap/sqlrootmap.go

```diff
@@ -91,60 +252,84 @@ func (rm *sqlRootmap) GetHistorical(
 	database string,
 	producer string,
 	topic string,
-) ([]RootListing, error) {
-	sql := `
-	select storage_prefix, node_id, version, timestamp
-	from rootmap
-	where database = ? and producer_id = ? and topic = ? 
-	order by version desc
-	`
-	rows, err := rm.db.QueryContext(ctx, sql, database, producer, topic)
+) (listings []RootListing, err error) {
+	rows, err := rm.db.QueryContext(ctx, `
+	select storage_prefix,
+	node_id,
+	rootmap.version,
+	rootmap.timestamp,
+	latest_truncations.version as truncation_version
+	from rootmap inner join tables on tables.id = rootmap.table_id
+	left join latest_truncations
+	on latest_truncations.table_id = tables.id
+	where tables.database = ? 
+	and tables.producer = ?
+	and tables.topic = ?
+	order by rootmap.version desc
+	`, database, producer, topic)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read from rootmap: %w", err)
+		return listings, fmt.Errorf("failed to read from rootmap: %w", err)
 	}
 	defer rows.Close()
-	listings := []RootListing{}
 	for rows.Next() {
 		var prefix, timestamp, nodeID string
+		var truncationVersion *uint64
 		var version uint64
-		if err := rows.Scan(&prefix, &nodeID, &version, &timestamp); err != nil {
-			return nil, fmt.Errorf("failed to read from rootmap: %w", err)
+		if err := rows.Scan(&prefix, &nodeID, &version, &timestamp, &truncationVersion); err != nil {
+			return listings, fmt.Errorf("failed to read from rootmap: %w", err)
+		}
+		minVersion := uint64(0)
+		if truncationVersion != nil {
+			if version <= *truncationVersion {
+				continue
+			}
+			minVersion = *truncationVersion + 1
 		}
 		decoded, err := hex.DecodeString(nodeID)
 		if err != nil {
-			return nil, fmt.Errorf("failed to decode node ID: %w", err)
+			return listings, fmt.Errorf("failed to decode node ID: %w", err)
 		}
 		listings = append(listings, RootListing{
-			prefix, producer, topic, nodestore.NodeID(decoded), version, timestamp, 0,
+			prefix, producer, topic, nodestore.NodeID(decoded), version, timestamp, minVersion,
 		})
 	}
 	if err := rows.Err(); err != nil {
-		return nil, fmt.Errorf("failed to read from rootmap: %w", err)
+		return listings, fmt.Errorf("failed to read from rootmap: %w", err)
 	}
 	return listings, nil
 }
 
-func (rm *sqlRootmap) GetLatestByTopic(
+func (rm *sqlRootmap) GetLatestByTopic( // nolint: funlen
```

Comment:
if you split out a separate function to build the SQL statement, do you still need the linting exception?

Reply:
Done!